### PR TITLE
Introduce ComputedUrl

### LINKED ISSUE #17625: Style computed value of a URL is returning the relative URL
If you call `style.computed_value_to_string(some-url-valued-property)` then you get the original, possibly relative, URL rather than the computed absolute URL. The culprit is https://github.com/servo/servo/blob/f18782fb0e10e150f44719d9899a62b84ecb7adc/components/style/servo/url.rs#L113-L121, which prioritizes the original URL, together with the fact that we share code between specified and computed URLs.

### PR DESCRIPTION
**Pull request by fnune**
Use the new `ComputedUrl` type for computed types and `SpecifiedUrl` for specified types instead of using the `SpecifiedUrl` implementation for both.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #17625 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because this is an implementation change and tests already exist.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17812)

### CODE REVIEW

**Comment by emilio** [comment]
Yeah, this is not the right patch and I bet it fails tests.

What needs to happen is that this piece of code needs to go away, at least for Servo:

```rust
// TODO(emilio): Maybe consider ComputedUrl to save a word in style structs?
impl ComputedValueAsSpecified for SpecifiedUrl {}
```

And instead we need to make something like:

```rust
enum ComputedUrl {
    Invalid(Arc<String>),
    Valid(ServoUrl),
}
```

And `impl ToComputedValue for SpecifiedUrl` properly, and also `ToCss`.

@bors-servo try

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /_mozilla/css/import_serialization.html:
  │ FAIL [expected PASS] CSS Test: @import rules are correctly serialized
  │   → assert_equals: expected "@import url(\"foo.css\");" but got "@import url(\"http://web-platform.test:8000/_mozilla/css/foo.css\");"
  │ 
  │ @http://web-platform.test:8000/_mozilla/css/import_serialization.html:13:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/_mozilla/css/import_serialization.html:11:1
```

**Comment by brainlessdeveloper** [comment]
Yep seems pretty clear. Would the new struct need to be used anywhere else?

**Comment by emilio** [comment]
I don't think so, for now at least.

**Comment by brainlessdeveloper** [comment]
Working on this now!

**Comment by brainlessdeveloper** [comment]
What should `SpecifiedUrl::to_css` return if there is no `original`? Because after these changes, `SpecifiedUrl` should not try to access `resolved`, right?

**Comment by brainlessdeveloper** [comment]
I'd like some input in the current progress. Also, currently I'm trying to understand some (to me) quite cryptic error messages during compilation, which I think are related to the associated type `ComputedValue` in the implementation of `ToComputedValue for SpecifiedUrl`:

```
error[E0308]: mismatched types                 
   --> /home/fausto/Dev/rust/servo/components/style/values/generics/image.rs:160:35
    |                                          
160 | #[derive(Clone, Debug, PartialEq, ToComputedValue, ToCss)]
    |                                   ^^^^^^^^^^^^^^^ expected struct `servo::url::SpecifiedUrl`, found enum `values::computed::ComputedUrl`
    |                                          
    = note: expected type `servo::url::SpecifiedUrl`                                          
               found type `values::computed::ComputedUrl`                                     

error[E0271]: type mismatch resolving `<servo::url::SpecifiedUrl as values::computed::ToComputedValue>::ComputedValue == servo::url::SpecifiedUrl`
   --> /home/fausto/Dev/rust/servo/components/style/values/generics/image.rs:160:35
    |                                          
160 | #[derive(Clone, Debug, PartialEq, ToComputedValue, ToCss)]
    |                                   ^^^^^^^^^^^^^^^ expected enum `values::computed::ComputedUrl`, found struct `servo::url::SpecifiedUrl`
    |                                          
    = note: expected type `values::computed::ComputedUrl`                                     
               found type `servo::url::SpecifiedUrl` 
```

Unfortunately I don't know how to go on about debugging those :c

**Comment by emilio** [comment]
> I'd like some input on the current progress. Also, currently I'm trying to understand some (to me) quite cryptic error messages during compilation, which I think are related to the associated type ComputedValue in the implementation of ToComputedValue for SpecifiedUrl:

Hi! The work so far looks quite reasonable :)

The errors you're seeing is because `values::generic::Image` is supposed to serve both as a computed image and a specified image, depending on the template parameters. That is, the `ToComputedValue` implementation does something like:

```rust
impl<Gradient, MozImageRect> ToComputedValue for Image<Gradient, MozImageRect>
where
    Gradient: ToComputedValue,
    MozImageRect: ToComputedValue,
{
    type ComputedValue = Image<Gradient::ComputedValue, MozImageRect::ComputedValue>;

    fn to_computed_value(&self, context: &Context) -> Self::ComputedValue {
        match *self {
            Image::Url(ref url) => Image::Url(url.to_computed_value(context)),
            Image::Gradient(ref gradient) => Image::Gradient(gradient.to_computed_value(context)),
            // ...
        }            
    }
}
```

As you may see, now that `url.to_computed_value(context)` returns something that isn't a `SpecifiedUrl`, the `Image::Url` value would be invalid.

Basically, you need to make `Image` generic also on the `Url` type, so it can be `SpecifiedUrl` when used as a specified value, and `ComputedUrl` when used as a computed value.

Hope it helps!

**Comment by emilio** [comment]
> What should SpecifiedUrl::to_css return if there is no original? Because after these changes, SpecifiedUrl should not try to access resolved, right?

Why not? We only won't have `specified` using the uncompute functionality, and in that case it's fine to just use the resolved

**Comment by brainlessdeveloper** [comment]
That does help a lot! Thanks. I'm making progress but there's a problem after some changes I've made to a file that uses templating (`.mako.rs`). After a failed build, a consecutive attempt will fail earlier and report some error in an `out` directory, and I have to remove the build folder and start again:
```
 servo git:(use-resolved-url-instead-of-original) ✗ ./mach build -d
error: failed to read `/home/fausto/Dev/rust/servo/target/debug/build/osmesa-src-61979e1c0efa04c9/output`

Caused by:
  No such file or directory (os error 2)
```
I can continue making changes like this but it's pretty annoying, am I missing anything?

**Comment by emilio** [comment]
So... That's pretty weird, seems like the osmesa build got messed up somehow...

You can probably delete the `osmesa-src-<hash>` directory, I'd expect that to restart the build from scratch, so it should work.

If that doesn't work, you can probably delete https://github.com/servo/servo/blob/master/ports/glutin/Cargo.toml#L42-L56, since osmesa is only there to provide the mesa runtime when running tests.

**Comment by brainlessdeveloper** [comment]
I see, I've had errors like that come up in different builds, for example right now in `target/debug/build/style-c1fe248c3e80ee66/out/properties.rs:51081:33`. I'll see if I can make progress faster with your recommendation, thanks.

**Comment by emilio** [comment]
Oh, yeah, the errors in `properties.rs` are annoying, but they come from the multiple `.mako.rs` files. Usually it's not hard to open the autogenerated file and track down which of the `longhands/foo.mako.rs` or `shorthands/foo.mako.rs` files it comes from.

**Comment by brainlessdeveloper** [comment]
So, I got it to compile but the tests aren't passing yet. Current status: I'm getting errors in files generated by `mako` templates. Example:
```
error[E0308]: mismatched types                                                
     --> /home/fausto/Dev/rust/servo/target/debug/build/style-9e8aa6a6aa746743/out/properties.rs:45265:66
      |                                                                               
45265 |                                 context.builder.set_marker_start(computed)                       
      |                                                                  ^^^^^^^^ expected struct `servo::url::SpecifiedUrl`, found enum `values::computed::ComputedUrl`
      |                                                                                                  
      = note: expected type `values::Either<servo::url::SpecifiedUrl, _>`                                                              
                 found type `values::Either<values::computed::ComputedUrl, _>`
```
I guess I should make structs like this:
```
<%helpers:shorthand name="marker" products="gecko"
    sub_properties="marker-start marker-end marker-mid"
    spec="https://www.w3.org/TR/SVG2/painting.html#MarkerShorthand">
    use values::specified::UrlOrNone;
```
...generic over the `Url` type. But I'm not too sure how bcause `UrlOrNone` is defined in the `specified` directory, implying it shouldn't be "specified/computed url or none".

Anyway, I'll be on this again tomorrow evening if I get some free time :)

**Comment by emilio** [comment]
That's because in `values/computed/mod.rs` there's a line that reads:

```rust
pub use super::specified::{BorderStyle, UrlOrNone};
```

But that's not true anymore, since `computed::UrlOrNone` shouldn't be just `specified::UrlOrNone`. So you'll have to remove `UrlOrNone` from there, and add something like:

```rust
pub type UrlOrNone = Either<ComputedUrl, None_>
```

I believe that should be enough to fix that compiler error.

Thanks again for working on this, it's great!

**Comment by brainlessdeveloper** [comment]
Previously failing `"CSS Test: @import rules are correctly serialized"` from WPT now passes but I haven't tested the entire suite. r? @emilio

**Comment by brainlessdeveloper** [comment]
This fails on `./mach build-geckolib`:

```
15 | use servo_url::ServoUrl;                              
   |     ^^^^^^^^^^^^^^^^^^^ Maybe a missing `extern crate servo_url;`? 
```

And exactly the same on `./mach test-stylo`. I suppose it's because of these `cfg`s:
```
#[cfg(feature = "servo")] #[macro_use] extern crate serde; 
#[cfg(feature = "servo")] extern crate servo_url;
```

Should I just remove the `feature` so it compiles for gecko too or should I do it some other way?

**Comment by emilio** [comment]
Hmm... I think we want ComputedValueAsSpecified for Gecko's URLs, for now, and ComputedUrl being a servo-only type. It's somewhat annoying though.

**Comment by brainlessdeveloper** [comment]
Alright I'll try to fix it later, thanks!

**Comment by brainlessdeveloper** [comment]
Alright, now everything passed on my machine, except for `wpt`, which I haven't run yet. @emilio r?

**Comment by brainlessdeveloper** [comment]
I don't know what to do about a failure on the Mac suite.

**Comment by KiChjang** [comment]
@bors-servo retry
* infra phail

**Comment by brainlessdeveloper** [comment]
r?

File: components/style/values/computed/mod.rs [hunk 1]
```diff
@@ -586,3 +589,45 @@ impl ClipRectOrAuto {
 
 /// <color> | auto
 pub type ColorOrAuto = Either<Color, Auto>;
+
+/// The computed value of a CSS `url()`, resolved relative to the stylesheet URL.
+#[cfg(feature = "servo")]
+#[derive(Clone, Debug, HeapSizeOf, Serialize, Deserialize, PartialEq)]
+pub enum ComputedUrl {
```
Comment (emilio) [inline components/style/values/computed/mod.rs]: We could move this to another module to avoid the `cfg`'s, but that's probably doable as a followup.

File: components/style/servo/url.rs [hunk 2]
```diff
@@ -126,3 +127,25 @@ impl ToCss for SpecifiedUrl {
...
     }
 }
+
+impl ToComputedValue for SpecifiedUrl {
+    type ComputedValue = ComputedUrl;
+
+    fn to_computed_value(&self, _: &Context) -> Self::ComputedValue {
+        match self.resolved {
+            Some(ref url) => ComputedUrl::Valid(url.clone()),
+            None => ComputedUrl::Invalid(Arc::new("about:invalid".to_string())),
```
Comment (emilio) [inline components/style/servo/url.rs]: Why `about:invalid`, and not the specified one? Off-hand, that should be both cheaper and more correct.

What do other browsers do here?

File: components/style/servo/url.rs [hunk 3]
```diff
@@ -126,3 +127,25 @@ impl ToCss for SpecifiedUrl {
...
+        }
+    }
+
+    fn from_computed_value(computed: &ComputedUrl) -> Self {
```
Comment (emilio) [inline components/style/servo/url.rs]: Also, if you did what I suggest above, you could use `original: None, resolved: Some()` for valid urls, and `original: Some(), resolved: None` for invalid ones, which would match what we do when computing them.

**Review by emilio — COMMENTED** [review]
Looks pretty good! Just some questions.

File: components/style/values/computed/mod.rs (hunk 1 above)
Comment (brainlessdeveloper) [inline components/style/values/computed/mod.rs]: You mean I should create a new files like this:
```
computed/ur/servo.rs
computed/url/gecko.rs
computed/url/mod.rs // Export here? Would still need the cfgs am I right?
```

File: components/style/servo/url.rs (hunk 2 above)
Comment (brainlessdeveloper) [inline components/style/servo/url.rs]: Sounds like a better idea.
I'll research other browsers this afternoon, I'll be in the Mozilla offices here in Berlin and I can probably ask directly :D

File: components/style/servo/url.rs (hunk 3 above)
Comment (brainlessdeveloper) [inline components/style/servo/url.rs]: Agree here too.

File: components/style/values/computed/mod.rs [hunk 4]
```diff
@@ -586,3 +589,45 @@ impl ClipRectOrAuto {
...
+        }
+    }
+}
+
+#[cfg(feature = "servo")]
+impl ToCss for ComputedUrl {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        let string = match *self {
+            ComputedUrl::Valid(ref url) => url.as_str(),
+            ComputedUrl::Invalid(ref invalid_string) => invalid_string,
```
Comment (brainlessdeveloper) [inline components/style/values/computed/mod.rs]: With the new changes this `invalid_string` should either be the original URL if it was specified, or `"about:invalid"` if not specified.

**Comment by brainlessdeveloper** [comment]
r?

File: components/style/servo/url.rs [hunk 5]
```diff
@@ -126,3 +127,33 @@ impl ToCss for SpecifiedUrl {
...
+    type ComputedValue = ComputedUrl;
+
+    // If we can't resolve the URL from the specified one, we fall back to the original
+    // but still return it as a ComputedUrl::Invalid
+    fn to_computed_value(&self, _: &Context) -> Self::ComputedValue {
+        match self.resolved {
+            Some(ref url) => ComputedUrl::Valid(url.clone()),
+            None => match self.original {
+                Some(ref url) => ComputedUrl::Invalid(url.clone()),
+                None => ComputedUrl::Invalid(Arc::new("about:invalid".to_string())),
```
Comment (emilio) [inline components/style/servo/url.rs]: I don't think this can be reached (and it'd be tricky because `about:invalid` is a valid URL. I think we should be able to mark this `unreachable!("Found specified url with neither resolved or original  URI!")`, or something like that.

File: components/style/values/computed/mod.rs (hunk 1 above)
Comment (emilio) [inline components/style/values/computed/mod.rs]: Yeah, I meant this, but I wouldn't worry about that for this PR, it's fine to do it in followup work.

**Review by emilio — APPROVED** [review]
Looks great, thanks!

Please double-check me in the "can't be reached" path, and let's get this merged! :)

File: components/style/servo/url.rs (hunk 5 above)
Comment (brainlessdeveloper) [inline components/style/servo/url.rs]: I think this is the same case as the `ToCss` implementation for `SpecifiedUrl`, when there's no original and no resolved:
```
// This can only happen if the url wasn't specified by the
// user *and* it's an invalid url that has been
// back to specified value via the "uncompute" functionality.
```
So I think the case is reachable, but I'm not sure the behavior is appropriate.

Comment (emilio) [inline components/style/servo/url.rs]: Right, but with your patch we're changing that, right? And the uncompute stuff (which is the `from_computed_value` implementation) now will always have one or the other.

**Comment by emilio** [comment]
Also, can we add a test for this, given none are written?

It'd be something like:

```
<div style="background-image: url(relative.png);"></div>
```

And assert that getComputedStyle(div).backgroundImage returns the resolved value, and div.style.backgroundImage the unresolved one.

File: components/style/servo/url.rs (hunk 5 above)
Comment (brainlessdeveloper) [inline components/style/servo/url.rs]: Yeap you're right. That was obvious... in hindsight :) I'll take care of it.

**Comment by brainlessdeveloper** [comment]
I wanted to write one like that but couldn't figure out where/how exactly. I've never written tests for servo. Where should I place it, and can I find any existing examples which are similar?

**Comment by emilio** [comment]
So you can write a new test as described in: https://github.com/servo/servo/blob/master/tests/wpt/README.md.

What you want is probably a very simple test setting the background-image of an element and checking `element.style.backgroundImage` and `getComputedStyle(element).backgroundImage` for differences.

Since it's hard to find a spec that backs it out, even though other browsers seem to do the same, we can put it in `tests/wpt/mozilla/tests/css`.

**Comment by brainlessdeveloper** [comment]
So, writing the tests was a good idea (duh). Apparently the `list-style-image` property and the shorthand `list-style` still use the specified URL. I'm onto fixing that.

Also, I can't manage to write tests for `content: url(test.jpg);` because I can't get `getComputedStyle(elem, "::after");` to return my styles. If I create a `div` with an `::after` and set `content: url(test.jpg);` on it, `getComputedStyle(elem, "::after");` will return `normal` which is not what I set. I haven't managed to work around this.

**Comment by brainlessdeveloper** [comment]
So, the trouble with `list-style-image` is [here](https://github.com/servo/servo/blob/4f0821192c112943bb53b4fb04303c1afdde06e6/components/style/properties/longhand/list.mako.rs#L103) but I can't figure out how to make the type generic for the URL... I'm getting really confused with the templating.

**Comment by brainlessdeveloper** [comment]
Sorry about the terrible force pushes, I messed up the rebase because it introduced compile errors.

**Comment by brainlessdeveloper** [comment]
So to fix the new WPT case I need to make the implementations [here](https://github.com/servo/servo/blob/4f0821192c112943bb53b4fb04303c1afdde06e6/components/style/properties/longhand/list.mako.rs#L103) generic on the URL type. Right now there's some code reuse going on with `ComputedValueAsSpecified`. The problem is I can't figure out how to make my changes in the template, I don't understand where I can introduce type parameters or whether I should rebuild the property somewhere else.

Any ideas?

**Comment by jdm** [comment]
You may need to introduce a [UrlOrNone](https://doc.servo.org/style/values/specified/type.UrlOrNone.html) type that uses computed URLs; you could then use that for the computed value, and use the original UrlOrNone type forthe specified value and remove the ComputedValueAsSpecified logic.

Note; these are guesses. If they sound sensible, give it a try while you're waiting for an answer from someone more knowledgeable ;)

**Comment by brainlessdeveloper** [comment]
Yep, I've already implemented a `UrlOrNone` type for computed URLs, but I think the problem is `parse` should be generic over the URL type and I can't add a type parameter to it because the `impl` block is implicit because of the template stuff (I think?).

**Comment by jdm** [comment]
When does `parse` need to return a computed URL value?

**Comment by jdm** [comment]
Keep in mind that the `pub fn parse` is just a function inside of a `list_style_image` module; it's not a method that's part of an impl.

**Comment by brainlessdeveloper** [comment]
I can't figure out how to make the tests pass. I don't understand how the template translates to the actual implementation.

On a side note, I just moved and I won't have an internet connection at home until next week, so I won't be able to work on this until then.

**Comment by jdm** [comment]
To see the actual implementation, open up target/debug/build/style-[hash]/out/properties.rs and search for `mod list_style_image`.

**Comment by brainlessdeveloper** [comment]
I won't be able to work during this week so I've [updated the test expectations for `listImage` properties](https://github.com/servo/servo/pull/17812/files#diff-b4e4af6248b356d099b97b18764bd91aR1). Is it fine if we try to merge this and I create an issue for the missing functionality?

The branch should pass the WPT as it is.

File: components/style/values/animated/mod.rs [hunk 6]
```diff
@@ -13,13 +13,15 @@ use smallvec::SmallVec;
...
 use values::computed::BorderCornerRadius as ComputedBorderCornerRadius;
+use values::computed::ComputedUrl;
 use values::computed::GreaterThanOrEqualToOneNumber as ComputedGreaterThanOrEqualToOneNumber;
 use values::computed::MaxLength as ComputedMaxLength;
 use values::computed::MozLength as ComputedMozLength;
 use values::computed::NonNegativeAu;
 use values::computed::NonNegativeLengthOrPercentage as ComputedNonNegativeLengthOrPercentage;
 use values::computed::NonNegativeNumber as ComputedNonNegativeNumber;
 use values::computed::PositiveInteger as ComputedPositiveInteger;
+#[cfg(feature = "servo")]
```
Comment (emilio) [inline components/style/values/animated/mod.rs]: You should remove this `cfg()`, afaict, for `./mach build-geckolib` to pass... or moving it to the `ComputedUrl`, probably.

**Review by emilio — COMMENTED** [review]
It looks reasonable to me, this is pretty neat work!

Let's:

 * File a followup issue (happy to do so, though you may have also a bit more context :)). Also, let's reference it from a comment near the relevant place.
 * Fix the build (which is failing in Travis, and which fix I think is a one-liner I suggest below)

And then  and get this landed, so it doesn't bitrot.

Thanks a lot again!

**Comment by brainlessdeveloper** [comment]
#18015

File: components/style/values/computed/mod.rs [hunk 7]
```diff
@@ -39,9 +43,8 @@ pub use self::image::{Gradient, GradientItem, Image, ImageLayer, LineDirection,
 pub use self::gecko::ScrollSnapPoint;
 pub use self::rect::LengthOrNumberRect;
 pub use super::{Auto, Either, None_};
-pub use super::specified::{BorderStyle, UrlOrNone};
+pub use super::specified::{BorderStyle};
```
Comment (emilio) [inline components/style/values/computed/mod.rs]: nit: I'm not sure if the lint catches this, but we usually don't wrap in braces single imports.

File: components/style/values/computed/mod.rs [hunk 8]
```diff
@@ -645,3 +648,45 @@ impl From<Au> for NonNegativeAu {
...
+            ComputedUrl::Invalid(ref invalid_string) => invalid_string,
+        };
+
+        dest.write_str("url(")?;
+        string.to_css(dest)?;
+        dest.write_str(")")
+    }
+}
+
+#[allow(missing_docs)]
```
Comment (emilio) [inline components/style/values/computed/mod.rs]: nit: could you use `<url> | <none>` as the doc comment?

File: tests/wpt/mozilla/meta/css/get-computed-style-for-url.html.ini [hunk 9]
```diff
@@ -0,0 +1,9 @@
+[get-computed-style-for-url.html]
+  type: testharness
+
+  [getComputedStyle(elem) for url() listStyle uses the resolved URL and elem.style uses the original URL]
+    expected: FAIL
```
Comment (emilio) [inline tests/wpt/mozilla/meta/css/get-computed-style-for-url.html.ini]: Let's add a `bug: https://github.com/servo/servo/issues/18015` line here and below.

File: tests/wpt/mozilla/tests/css/get-computed-style-for-url.html [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id="tests"></div>
+<script>
+  var origin = window.location.origin;
+  var container = document.getElementById("tests");
+
+  // Some shorthands like background and border-image compute to other properties
+  var computedPropertyName = {
+    background: "backgroundImage",
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/get-computed-style-for-url.html]: You could just set `backgroundImage`, etc instead... though not a big deal.

File: tests/wpt/mozilla/tests/css/get-computed-style-for-url.html [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+  function makeComputedUrlStyle(path) {
+    return `url("${origin}/${path}")`;
+  }
+  function makeSpecifiedUrlStyle(path) {
+    return `url("${path}")`;
+  }
+  function testUrlsForProperty(property, urlValue, extraShorthand) {
+    test(function() {
+      var extra = extraShorthand ? ` ${extraShorthand}` : "";
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/get-computed-style-for-url.html]: nit: `extraShorthand ? extraShorthand.toString() : ""`?

Comment (emilio) [inline tests/wpt/mozilla/tests/css/get-computed-style-for-url.html]: (But setting the longhand property directly would allow you to avoid this complexity).

**Review by emilio — APPROVED** [review]
r=me. I think the test could be simpler, but it's probably not a huge deal I guess? (it'll surely be nicer though)

File: tests/wpt/mozilla/tests/css/get-computed-style-for-url.html (hunk 11 above)
Comment (brainlessdeveloper) [inline tests/wpt/mozilla/tests/css/get-computed-style-for-url.html]: The problem is the space before the `extraShorthand`, which I add using the template string. It's not too nice...

File: tests/wpt/mozilla/tests/css/get-computed-style-for-url.html (hunk 10 above)
Comment (brainlessdeveloper) [inline tests/wpt/mozilla/tests/css/get-computed-style-for-url.html]: Yeap, I'm doing that as well, but I wanted to test the shorthand and the longhand too.

**Comment by brainlessdeveloper** [comment]
r?

**Comment by emilio** [comment]
@bors-servo r+

Thanks!
